libraries/chain/include/eosio/chain/thread_utils.hpp

```diff
@@ -12,6 +12,74 @@
 
 namespace eosio { namespace chain {
 
+   // should be defined for c++17, but clang++16 still has not implemented it
+#ifdef __cpp_lib_hardware_interference_size
+   using std::hardware_constructive_interference_size;
+   using std::hardware_destructive_interference_size;
+#else
+   // 64 bytes on x86-64 │ L1_CACHE_BYTES │ L1_CACHE_SHIFT │ __cacheline_aligned │ ...
+   [[maybe_unused]] constexpr std::size_t hardware_constructive_interference_size = 64;
+   [[maybe_unused]] constexpr std::size_t hardware_destructive_interference_size  = 64;
+#endif
+
+   // Use instead of std::atomic when std::atomic does not support type
+   template <typename T>
+   class large_atomic {
+      alignas(hardware_destructive_interference_size)
+      mutable std::mutex mtx;
+      T value{};
+   public:
+      T load() const {
+         std::lock_guard g(mtx);
+         return value;
+      }
+      void store(const T& v) {
+         std::lock_guard g(mtx);
+         value = v;
+      }
+
+      class accessor {
+         std::lock_guard<std::mutex> g;
+         T& v;
+      public:
+         accessor(std::mutex& m, T& v)
+            : g(m), v(v) {}
+         T& value() { return v; }
+      };
```

Comment:
This seems like a potentially dangerous design. I could do:

`auto& id = if_irreversible_block_id.make_accessor().value();`

and then access `id` outside of the lock.

I think if we use it only for `digest_type`, it would be preferable to call `load` and copy the 32 bytes, it is no big deal.

Reply:
It is not about avoiding the copy. It is to avoid `if_irreversible_block_id` being set to a lower block number.

```c++
   void set_if_irreversible_block_id(const block_id_type& id) {
      const block_num_type id_num = block_header::num_from_id(id);
      const bock_id_type current_lib_id = if_irreversible_block_id.load();
      const block_num_type current_num = block_header::num_from_id(current_lib_id);
      // thread 1 is setting LIB from 3 to 7
      // thread 2 is setting LIB from 3 to 5
      if( id_num > current_num ) {
         // thread 2: which is setting LIB to 5 pauses here
         // thread 1: which is setting LIB to 7 comes in and sets to 7
         // thread 1: unpauses and sets LIB back to 5
         dlog("set irreversible block ${bn}: ${id}, old ${obn}: ${oid}", ("bn", id_num)("id", id)("obn", current_num)("oid", accessor.value()));
         if_irreversible_block_id.store(id);
      }
   }



```

Reply:
Got it, makes sense. I think I would have preferred a custom struct, with a mutex, and a member function `advance_lib(block_num_type block_num);` which would do it safely.